### PR TITLE
feat: Bootstrap metastore for wal segments

### PR DESCRIPTION
**Pull request by cyriltovena**
**What this PR does / why we need it**:

This adds a new component `metastore` that is using raft for reliability and index the last 15m of wal segment ID.

This is very hacky and experimental to move forward with querying the wal segment without listing them all. This is going to change a lot over time.

TODO:

- Needs to wire that up in the write path now.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: docs/sources/shared/configuration.md [hunk 1]
```diff
@@ -2625,6 +2652,8 @@ The `frontend_worker` configures the worker - running within the Loki querier -
 
 # Configures the querier gRPC client used to communicate with the
 # query-scheduler. This can't be used in conjunction with 'grpc_client_config'.
+# The CLI flags prefix for this block configuration is:
```
Comment (cyriltovena) [inline docs/sources/shared/configuration.md:2655]: weird

File: pkg/storage/bloom/v1/bloom_tokenizer_test.go [hunk 2]
```diff
@@ -11,13 +11,14 @@ import (
...
 	"github.com/grafana/dskit/multierror"
 
-	"github.com/grafana/loki/pkg/push"
 	"github.com/grafana/loki/v3/pkg/chunkenc"
 	"github.com/grafana/loki/v3/pkg/iter"
 	v2 "github.com/grafana/loki/v3/pkg/iter/v2"
 	"github.com/grafana/loki/v3/pkg/logproto"
 	"github.com/grafana/loki/v3/pkg/logql/log"
 
+	"github.com/grafana/loki/pkg/push"
```
Comment (cyriltovena) [inline pkg/storage/bloom/v1/bloom_tokenizer_test.go:20]: make format got me this not sure why it wasn't already like that

File: pkg/ingester-rf1/flush.go [hunk 3]
```diff
@@ -115,10 +116,17 @@ func (i *Ingester) flushSegment(ctx context.Context, j int, w *wal.SegmentWriter
 	}
 
 	i.metrics.flushesTotal.Add(1)
-	if err := i.store.PutObject(ctx, fmt.Sprintf("loki-v2/wal/anon/"+id.String()), buf); err != nil {
+	if err := i.store.PutObject(ctx, fmt.Sprintf("loki-v2/wal/anon/"+id), buf); err != nil {
 		i.metrics.flushFailuresTotal.Inc()
 		return fmt.Errorf("store put chunk: %w", err)
 	}
 
+	if _, err := i.metastoreClient.AddBlock(ctx, &metastorepb.AddBlockRequest{
```
Comment (grobinson-grafana) [inline pkg/ingester-rf1/flush.go:124]: Do we need a kind of 2pc here? What do we do about orphaned objects that were written to object storage but for whatever reason couldn't be added to the metastore?

File: pkg/storage/bloom/v1/bloom_tokenizer.go [hunk 4]
```diff
@@ -7,12 +7,13 @@ import (
...
 	"github.com/go-kit/log/level"
 
-	"github.com/grafana/loki/pkg/push"
 	"github.com/grafana/loki/v3/pkg/iter"
 	v2iter "github.com/grafana/loki/v3/pkg/iter/v2"
 	"github.com/grafana/loki/v3/pkg/logproto"
 	"github.com/grafana/loki/v3/pkg/storage/bloom/v1/filter"
 	"github.com/grafana/loki/v3/pkg/util/encoding"
+
+	"github.com/grafana/loki/pkg/push"
```
Comment (grobinson-grafana) [inline pkg/storage/bloom/v1/bloom_tokenizer.go:16]: Fix?

**Review by grobinson-grafana — COMMENTED** [review]
Looks OK to me, however I didn't look at the metastore code in depth, just at the integration with ingester-rf1. If you would like a more in-depth review of the metastore code please ask!

File: pkg/ingester-rf1/flush.go (hunk 3 above)
Comment (cyriltovena) [reply pkg/ingester-rf1/flush.go:124]: I suspect they will get cleaned up by a something else.

File: pkg/storage/bloom/v1/bloom_tokenizer.go (hunk 4 above)
Comment (cyriltovena) [reply pkg/storage/bloom/v1/bloom_tokenizer.go]: lol I failed :)

File: docs/sources/shared/configuration.md (hunk 1 above)
Comment (cyriltovena) [reply docs/sources/shared/configuration.md:2628]: @jdbaldry something is off, I can't understand why this change is here ?

**Review by grobinson-grafana — APPROVED** [review]

**Pull request merged**